transaction/core/src/ring_signature/rct_bulletproofs.rs

```diff
@@ -50,6 +101,22 @@ pub struct OutputSecret {
     pub blinding: Scalar,
 }
 
+/// The parts of a TxIn needed to validate a corresponding MLSAG
+#[derive(Clone, Debug)]
+pub struct SignedInputRing {
+    /// A reduced representation of the TxOut's in the ring. For each ring
+    /// member we have only:
+    /// * The onetime-address (tx_out.target_key)
+    /// * The compressed commitment (tx_out.amount.commitment)
+    pub members: Vec<(CompressedRistrettoPublic, CompressedCommitment)>,
+
+    /// The digest that this signature is supposed to have signed.
+    /// If omitted, then the entire extended message digest should have been
+    /// signed. If the input has rules, then a reduced digest is signed
+    /// instead. (See MCIP #31 for rationale)
+    pub signed_digest: Option<[u8; 32]>,
```

Comment:
this was changed since last review from `Option<InputRule>` to `Option<[u8; 32]>`, so, if the input signs something other than the extended message digest, the thing it is supposed to sign instead goes here. That makes it easier to do a digest of the whole `TxIn` without bringing the whole `TxIn` into this scope.